virtual_rainforest/core/config.py

```diff
@@ -0,0 +1,362 @@
+            try:
+                toml_dict = tomllib.load(f)
+                # Check for repeated entries across previous nested dictionaries
+                for item in file_data:
+                    repeats = check_dict_leaves(item[1], toml_dict, [])
+                    for elem in repeats:
+                        conflicts.append((elem, file, item[0]))
+
+                file_data.append((file, toml_dict))
+            except tomllib.TOMLDecodeError as err:
+                log_and_raise(
+                    f"Configuration file {file} is incorrectly formatted. Failed with "
+                    f"the following message:\n{err}",
+                    RuntimeError,
+                )
+
+    # Check if any tags are repeated across files
+    if len(conflicts) != 0:
+        # If so generate full list of errors
+        msg = "The following tags are defined in multiple config files:\n"
+        for conf in conflicts:
+            msg += f"{conf[0]} defined in both {conf[1]} and {conf[2]}\n"
+        msg = msg[:-1]
+        log_and_raise(msg, RuntimeError)
+
+    # Merge everything into a single dictionary
+    config_dict: dict = {}
+    for item in file_data:
+        dpath.util.merge(config_dict, item[1])
+
+    return config_dict
+
+
+def find_schema(config_dict: dict) -> list[str]:
+    """Find which schema the configuration requires to be loaded.
+
+    Args:
+        config_dict: The complete configuration settings for the particular model
+            instance
+    """
+
+    # Find which other schema should be searched for
+    try:
+        modules = config_dict["core"]["modules"]
+    except KeyError:
+        log_and_raise(
+            "Core configuration does not specify which other modules should be "
+            "configured!",
+            KeyError,
+        )
+
+    # Add core to list of modules if its not already included
+    if "core" not in modules:
+        modules.append("core")
+
+    if len(modules) != len(set(modules)):
+        log_and_raise(
+            f"The list of modules to configure given in the core configuration file "
+            f"repeats {len(modules) - len(set(modules))} names!",
+            RuntimeError,
+        )
+
+    return modules
+
+
+def construct_combined_schema(modules: list[str]) -> dict:
+    """Load validation schema for desired modules, and combine into a single schema.
+
+    Args:
+        modules: List of modules to load schema for
+    """
+
+    # Construct combined schema for all relevant modules
+    comb_schema: dict = {}
+
+    # Loop over expected modules and add them to the registry
+    for module in modules:
+        if module in SCHEMA_REGISTRY:
+            try:
+                m_schema = SCHEMA_REGISTRY[module]["properties"][module]
+            except KeyError as err:
+                log_and_raise(
+                    f"Schema for {module} module incorrectly structured, {err} key "
+                    f"missing!",
+                    KeyError,
+                )
```

Comment:
If schemas are checked when added to the registry, then this is not needed, rigth?

Reply:
The schema checking when adding to the registry just checks whether the schema is a valid JSON schema. This catches a bit more, i.e. whether the schema contains the keys we require which could be missing from an otherwise valid schema. I could move this check to the schema registry decorator if that's a better place for it?

Reply:
I think it makes sense. If the schema MUST have those keys, then the schema is just not valid without them, so it should not be registered to start with, right?